### ISSUE TITLE
Extend Python FFI bindings to adopt more languages

### ISSUE THREAD

**Issue by Pzixel** [opencv/opencv#11666]
It's not a bug report, but it's not a question too, so I think it's an appropriate place to write this request.

There is plenty of projects that are trying to implement FFI bindings to opencv, e.g. [this](https://github.com/hybridgroup/gocv) or [this](https://github.com/ruby-opencv/ruby-opencv) among many others. And they are just doing the same things over and over again.

But this repository has already great `gen2.py` that consider all edge cases and may be used to generate all required wrappers. But currently it requires some preinstalled software to be ran (e.g. numpy) as well as it generates python code which makes it hard to use for anything but python.

I have two proposals, the first one is easy to implement, the second is a bit more complicated, but has a great value

1. Create some target that would only generate python wrappers. It shouldn't check any environment dependencies, the only thing it should do is generate some code. That would allow other languages to use this code as an intermediate representation to generate their own wrappers
1. Make `gen2.py` (implement some new mechanism?) to generate more appropriate MIR (middle-level IR) and then generate python bindings from this representation. It requires more effort but it leads to great simplification for wrapper authors as well as it may be considered to be some kind of doogfooding where python bindings uses ther same MIR as others. It guarantees any FFI built on top of that would be full-featured and bugless, with all edge cases covered. The community effort would be focused in one place instead of being dissipated among multiple repositories trying to solve same issues over and over again

This would lead to better language adoption, more contributions, and more spreading overall.

I do believe it's the right way to do it and it's the high priority task for the repo. Let me provide a @berak 's  quote from our conversation:

> adapting a ~1000 lines python monster (which to 90% covers weird edge cases) might look daunting, but on the other hand, doing things [manually, like those folks](https://github.com/hybridgroup/gocv/) try will result in a "toy example", that only covers the 5% most used cases. (which again IS cool, too, don't get me wrong. problem is, if you deviate only a millimeter from it, you're staring into an open abyss)

We currently only have python and C++ code that works, and plenty of "toy examples" in other languages. I think it shall be changed.

**Comment by alalek** [comment]
Related comment about C FFI interface: https://github.com/opencv/opencv/issues/11000#issuecomment-376683995

C FFI generator may provide some meta-information about handled classes / functions / parameters to support adoption of other languages.

**Comment by Pzixel** [comment]
Yes, C could be this intermediate representation (or maybe some other type if convinient), what a great idea. I had initially an idea to wrap over C api but they said that it's obsolete for a long long time.

P.S. I wonder why this discussion is so sluggish. Nobody is interested in?

**Comment by StevenPuttemans** [comment]
> P.S. I wonder why this discussion is so sluggish. Nobody is interested in?

It has been 3 days, and you got respons in a day, I guess that is quite fast, compared to other issues :)

**Comment by Pzixel** [comment]
@StevenPuttemans yeah, @alalek was incredibly fast, but he's the only who did it.

I can provide any help, if required. But this task seems to be too big for one person.

**Comment by jbruening** [comment]
there are also a couple of csharp projects that have partially generated/hand written c bindings

https://github.com/shimat/opencvsharp
https://github.com/emgucv/emgucv

**Comment by Pzixel** [comment]
@StevenPuttemans and could you agree now that discussion is stalled? :)

I can provide any help required to accomplish this, but I just don't know how to do it. I'm not a C or C++ dev, so my knowledge is quite limited, but I'd really like OpenCV to be the only focus point where all CV algorithms gets implemented, and then they get available in all the languages on all the platforms.

**Comment by StevenPuttemans** [comment]
> @StevenPuttemans and could you agree now that discussion is stalled? :)
Ah well, to me it just shows that although your point raised is valid, it is not on the priority list of OpenCV. Tags are added so your request is added into the cue, but there is no telling when they will pick if you do not pick it up yourself :D Or start a campaign to find people that want to help.

**Comment by Pzixel** [comment]
Any possibility it will be a priority? Bringing opencv to the more platforms should be very important, I just don't get why it has priority so low...

**Comment by StevenPuttemans** [comment]
@Pzixel as said before, if you want to get things done in OSS communities, then you have to gather a club and get your hands dirty. There are tons of things on the to do list of this project, and this is just one of them. 

On the other hand, the call for GSoC 2019 is open, you might be interested in submitting it as a proposal there?

**Comment by Pzixel** [comment]
> as said before, if you want to get things done in OSS communities, then you have to gather a club and get your hands dirty. There are tons of things on the to do list of this project, and this is just one of them.

I would be glad to do it, but I don't know neither Python or C++. I'm a C# dev and I just don't have required qualification. I did my best in trying to wrap C++ code manually, but I can't read thousands of python codegen and make files to get it work. It just too hard to me. So I just hope someone who actually does know them will take a look on it. I'm not demanding anything, I'm asking for.

> On the other hand, the call for GSoC 2019 is open, you might be interested in submitting it as a proposal there?

Yes, writing proposal is in my powers.

**Comment by Pzixel** [comment]
Hmm, it looks like it's a conference (you are talking about google summer, right?) so I don't know how to send a proposal there.

**Comment by StevenPuttemans** [comment]
Google summer of code is a google summer project from May to Augusts that pays promising students to work on open source software during the summer. OpenCV as an organization applied again this year and got accepted. Now we are in the process of streamlining everything

https://github.com/opencv/opencv/wiki/GSoC_2019

More info on the link above, but your best bet would be to propose to mentor and get your topic out while the proposals are still flowing in!

**Comment by Pzixel** [comment]
Hello again.

I finally managed to send a request, but I only found a mentoring form. I have an idea which may get implemented, but I'm not a skilled C++ dev to mentor anyone. I'm pretty good in C# and know some Rust, I only know C++ to be able to create some hand-written wrappers. How can I throw a ticked into idea bucket?

**Comment by alalek** [comment]
Current bindings generators for Python/Java/DoxygenPythonEntries are written on Python.

We prefer to avoid "hand-written wrappers".

**Comment by Pzixel** [comment]
Me as well. But everybody has to write its own generators from C++ which has no stable ABI, has mangling instability and so on.

C is lungua franca so generating C from C++ first, and then generate Python/Java/... from C would benefit everyone since C is stable. There is `rust bindgen` for example which is an existing tool that may generate the whole opencv wrapper. You only have to write some ideomatic wrappers on top of that and that's all. But it can't resilently generate bindings from C++ code, or python, or anything. So we have to write our own generators (see linked issue), face the same issues over and over again in Ruby/C#/Rust/Go/Lua/JS/younameit. It's just a repetitive work people do again and again, facing the same cumbersomes and so on. I'd really like to change something. But unfortunately I'm not a C++ dev neither Python one, I can't analyze the whole `gen2.py` and write my own which would generate some C compatible code. I can only ask and propose here.

**Comment by Pzixel** [comment]
So? I'm really stuck. I think there is no much time left to leave a suggestion for the coding summer.

**Comment by StevenPuttemans** [comment]
I am sorry, but I am currently to busy to continue pitching in here.
